pkg/azdiskschedulerextender/azdiskschedulerextender.go

```diff
@@ -354,21 +335,28 @@ func getKubernetesClientset() (*kubernetes.Clientset, error) {
 	return kubeClient, nil
 }
 
-func getNodeScore(replicaCount int, attachingReplicaCount int, latestHeartbeat metav1.Time, availableNodeCount int, totalVolumeAttachmentCount int, nodeName string) int64 {
+func getNodeScore(createdAttachmentsForPodCount, attachingAttachmentsForPodCount, allAttachmentsCount int, latestHeartbeat metav1.Time, node *v1.Node) int64 {
 	now := time.Now()
 	latestHeartbeatWas := latestHeartbeat.Time
 	latestHeartbeatCanBe := now.Add(-2 * time.Minute)
 
 	if latestHeartbeatWas.Before(latestHeartbeatCanBe) {
 		klog.Warningf(
 			"node is unresponsive. Latest node heartbeat for %v was: %v. Latest accepted heartbeat is: %s",
-			nodeName,
+			node.Name,
 			latestHeartbeatWas.Format(time.UnixDate),
 			latestHeartbeatCanBe.Format(time.UnixDate))
 		return 0
 	}
 
-	score := int64(replicaCount*100) - int64(attachingReplicaCount*25) - int64((totalVolumeAttachmentCount-replicaCount)*10)
+	var remainingNodeCapacity int
+	if nodeMaxCapacity, err := azureutils.GetNodeMaxDiskCountWithLabels(node.Labels); err != nil {
+		remainingNodeCapacity = 0
+	} else {
+		remainingNodeCapacity = nodeMaxCapacity - allAttachmentsCount
+	}
+
+	score := int64(createdAttachmentsForPodCount*100) + (100 - int64(attachingAttachmentsForPodCount*10)) + int64(remainingNodeCapacity)
```

Comment:
Is there a reason we are updating the scoring logic?

Reply:
It's a small update to account for node's remaining disk count.

Reply:
`createdAttachmentsForPodCount`  with the current logic, already included `attachingAttachmentsForPodCount` so basically our formula now is something like: 
`(x+y)*100 +100-y*10+z = 100x+90y+z+100

Reply:
So, nodes that have more attachments in `Attaching` or `Failed` state will score higher with this logic.

Reply:
We could also count only `Attached` volumes in `createdAttachmentsForPodCount`

Reply:
>So, nodes that have more attachments in Attaching state will score higher with this logic.

It's `(100 - (attaching * 10)`, so we are counting number of attached attachments here and prioritizing nodes with more requested disks attached.

Reply:
what I am saying is attaching is counted twice and when you simplify the formula, you get what I have.

Reply:
> what I am saying is attaching is counted twice and when you simplify the formula, you get what I have.

Yes, but you see that in the equation `x = number of requested attachments that's attached` and `y = number of requested attachments that's not yet attached`.

So, we give higher coefficient for attached attachments, and a slightly lower coefficient for attaching / pending attachments.

Reply:
x includes y is what I am saying. line 273. So you end up with `100x+90y+z+100` this is not what we want

Reply:
Actually, let me rephrase that - it is close to what we want but wouldn't it be better to have a bigger difference 100 vs 90 and I don't see the point of keeping the 100 constant

Reply:
I think we can do something like this: 
`score := int64(createdAttachmentsForPodCount*100) - int64(attachingAttachmentsForPodCount*50) + int64(remainingNodeCapacity)`
This will always be positive and attached volumes will be more strongly prioritized

Reply:
The subtraction from 100 was so that we can keep the score in a `xyz` fashion where each digit is kind of reserved for each priority criteria.  `x`, which is in the 100th place, to record score for mostly highly prioritized requirement, `y`, in the 10th place, for the second most highly prioritized requirement, and so on. I understand that this will not always be the case because `x`, `y`, `z` will not always be less than 10 but it does keep our priorities straight.

Reply:
> I think we can do something like this: `score := int64(createdAttachmentsForPodCount*100) - int64(attachingAttachmentsForPodCount*50) + int64(remainingNodeCapacity)` This will always be positive and attached volumes will be more strongly prioritized

This is then equating 1 attachment in attached state with 2 attachments in attaching state. Say there are 2 nodes, A and B. A has 2 attachments: 0 attached and 2 attaching. B has 1 attachment: 1 attached and 0 attaching. We end up giving same scores for the two nodes and worse, if A has less remaining node capacity than B, we will end up giving higher score to B, which is not what we want.

Reply:
Given that we have a clear ordering of preference in prioritization requirements, I think we should have our score should reflect on this gradation between each requirement properly.

Reply:
In your example it will be: 3*100-2*50 =200 vs 2*100-2*50=0, we will not be really equalizing the scores but yeah to avoid them being multiples, we can pick another number instead of 50.

Reply:
And I still think we should keep the baseline at 0 not 100.

Reply:
What I am saying is let's reserve each digit for individual requirement. 100s for score for number of _created_ attachments, 10s for score number of _attached_ attachments, and 1s for score for number of remaining node capacity.

Reply:
We could make this clearer by using `10 * number of attached attachments` instead of `100 - 10 * number of attaching attachments`

Reply:
> In your example it will be: `3*100-2*50 =200` vs `2*100-2*50=0`, we will not be really equalizing the scores but yeah to avoid them being multiples, we can pick another number instead of 50.

In the example, node A would have `2 * 100 - 2 * 50 = 100` and B would have `1 * 100 - 0 * 50 = 100`. So these 2 nodes would end up with the same score.

Reply:
line 273 you count all volumes regardless of status in createdAttachmentsForPodCount, line 275 those with Attaching state get counted **again** in  nodeNameToNumAttachingAttachmentsForPodsMap.

Reply:
`createdAttachmentsForPodCount` already includes `nodeNameToNumAttachingAttachmentsForPodsMap`

Reply:
Yes, I am aware as I commented above.

> > what I am saying is attaching is counted twice and when you simplify the formula, you get what I have.
> 
> Yes, but you see that in the equation `x = number of requested attachments that's attached` and `y = number of requested attachments that's not yet attached`.
> 
> So, we give higher coefficient for attached attachments, and a slightly lower coefficient for attaching / pending attachments.

Reply:
What I am saying is we can make the prioritization more discrete by keeping them in separate digits. We prioritize nodes based on the following order: 1) number of requested attachments created on the node, 2) number of requested attachments attached on the node, and 3) number of remaining node capacity.

And we should **always** prioritize node with more requested attachments over nodes with less created attachments even if the other node has its attachments in _attached_ state. And we can keep this clear ordering by putting those 3 scores in separate digits.

Reply:
:D

Reply:
Yes, that is what I am also saying and the right formula for calculating that score is: 
`core := int64(createdAttachmentsForPodCount*100) - int64(attachingAttachmentsForPodCount*50) + int64(remainingNodeCapacity)` and not what you have because `createdAttachmentsForPodCount=x+y` and `attachingAttachmentsForPodCoun=y`

Reply:
`score := int64(createdAttachmentsForPodCount*100) - int64(attachingAttachmentsForPodCount*50) + int64(remainingNodeCapacity)` will not guarantee this ordering as described above. We equate node A with 2 attachments (0 attached, 2 attaching) and B with 1 attachment (1 attached), as they will both end up with `100`.

Reply:
And with the score you suggest the 2 attaching will score higher than 1 attached 180 vs 100.

Reply:
And whatever coefficient you give for `attachingAttachmentsForPodCount`, it's not going to be any different. Say we give `30` instead of `50`.

Node A: `4 attachments (0 attached, 4 attaching)` will have `4 * 100 - 4 * 30= 180`
Node B: `2 attachments (2 attached, 0 attaching)` will have `2 * 100 - 0 * 30 = 200`

We end up giving higher score to B when we should be giving higher score to A.

Reply:
and like I said, you can make 50 to 25 where 1 attachment will be equivalent to to 4 attaching, which is what we had before

Reply:
I think `100 - attachingAttachmentsForPodCount * 10` threw you off. To keep things less confusing and more consistent, why don't we use `attachedAttachmentsForPodCount` instead of `attachingAttachmentsForPodCount`. And make the scoring formula more straightforward by replacing `100 - attachingAttachmentsForPodCount * 10` with `10 * attachedAttachmentsForPodCount`?

i.e. `score := 100 * createdAttachmentsForPodCount + 10 * attachedAttachmentsForPodCount + remainingNodeCapacity`

Reply:
Ok, so why do we want to give higher score to A? This is something new

Reply:
We want to give higher score to A because if we schedule pod on B, we have higher risk of having to detach and reattach volume in case the volume's `maxShares` is saturated.

Reply:
Always scheduling pod to node with most requested attachments minimizes this risk.

Reply:
I don't think anything was throwing me off and this is almost exactly what I am trying to get to `100 - attachingAttachmentsForPodCount * 10` doesn't have the intend that you wanted because it was giving a higher coefficient to attaching volumes than what we wanted. If you don't want to penalize for volumes being in attaching state but that is counted as advantage that's a different discussion and you can keep it as addition instead of the penalty like our previous logic was.

Reply:
I guess what I also don't understand is why you want to prioritize attaching volumes more than the attached ones. This workflow is exercised on failover and in most cases, we would have attached volumes on some node while there might be another node trying to attach them to new node because the initial node with primaries failed.

Reply:
I am not trying to prioritize attaching volumes over attached ones. That's why I suggested
`score := 100 * createdAttachmentsForPodCount + 10 * attachedAttachmentsForPodCount + remainingNodeCapacity`

Reply:
With the addition of logic in `CrdProvisioner` to detach replicas to host primaries, we are now more likely to break the replica grouping. And that is why we need to enclose with scheduleextender logic to minimize the risk of scheduling pod onto a node where we have to detach and reattach more volumes.

Reply:
OK, yes this is exactly what I am trying to say as well, except with attaching volumes.

However, with this formula we have 110*attached+100*attaching and with the formula initially you had 100*attached+90*attaching and I was wondering if we should keep this ratio bigger that is why I suggested 50 but replace 50 with 10 and you get the same thing as your formula but simpler.  `score := 100 * createdAttachmentsForPodCount - 10 * attaching + remainingNodeCapacity` you still get `100*attached + 90*attaching+remainingNodeCapacity`. This is all I am trying to get at.

Reply:
Yeah, I think we are in agreement. With `110*attached+100*attaching`, we will never end up scheduling pod to a node with less requested attachment and when the two nodes have the same number of requested attachments, this will allow us to differentiate between the two because we give higher coefficient to `attached`. 

When we penalize, no matter what coefficient we give for penalty, we could theoretically end up scheduling pod on node with less attachments, and that's all I am trying to get at :D

Reply:
Yes, you are right with the new logic penalizing for being in `attaching` state is no longer for our advantage

Reply:
And hopefully this is more readable to avoid such lengthy discussions in the future :D

Reply:
Will do. I will add comment to describe above as well.

Reply:
There wasn't much logic ourside of this function to update anyway. I have updated `int64(attachedAttachmentsForPodCount * 10)` instead for consistency and clarity.